### PR TITLE
Feat/user registration

### PR DESCRIPTION
**Pull request by MuhammedMagdyy**
##### Why do we need this change?
  - Added a new column into the database (photoURL), also created migrations.
  - Added Zod error handler into error handling middleware to handle validation errors.
  - Added validation inputs for both creating and updating posts.
  - Added uploading an image while the registration process, also added validation for creating a new user.

### CODE REVIEW

File: src/database/prisma/schema.prisma [hunk 1]
```diff
@@ -11,22 +11,23 @@ datasource db {
...
   preferredFoods   String?
   hobbies          String?
-  Role             Role     @default(USER)
+  Role             Role      @default(USER)
   verificationCode String?
-  isVerified       Boolean  @default(false)
-  createdAt        DateTime @default(now())
-  updatedAt        DateTime @updatedAt
+  isVerified       Boolean   @default(false)
+  photoURL         String
```
Comment (mohamadelhinamy) [inline src/database/prisma/schema.prisma:28]: I think the photo should be optional dont you think?

File: src/modules/Post/services/postService.ts [hunk 2]
```diff
@@ -1,10 +1,25 @@
...
+    })
+    .required();
+
+  updatePostSchema = z.object({
+    description: z.string().nonempty({
+      message: 'Description cannot be empty',
+    }),
+  });
+
+  async createOne(args: z.infer<typeof this.createPostSchema>) {
```
Comment (kerolloz) [inline src/modules/Post/services/postService.ts:22]: Why did you change the args type here?

File: src/modules/User/services/userService.ts [hunk 3]
```diff
@@ -4,11 +4,26 @@ import sendGridEmail from '../../../mailers/sendEmail';
...
+        .min(3, { message: 'First name is too short' })
+        .trim(),
+      lastName: z.string().min(3, { message: 'Last name is too short' }).trim(),
+      email: z.string().email({ message: 'Email must be unique' }).trim(),
+      password: z.string().min(6, { message: 'Password is too short' }).trim(),
+      photoURL: z.string().optional(),
+    })
+    .required();
+
+  async createOne(args: z.infer<typeof this.createUserSchema>) {
```
Comment (kerolloz) [inline src/modules/User/services/userService.ts:26]: same here

File: src/routes/postRouter.ts [hunk 4]
```diff
@@ -25,12 +25,14 @@ router.post(
...
-      photoURL,
+    const userId = Number(req.body.userId); // TODO: change to req.user.id after adding token
+    const postSchema = postService.createPostSchema.parse({
+      ...req.body,
       userId,
+      photoURL,
+    });
+    const newpostObject = await postService.createOne({
+      ...postSchema,
     });
```
Comment (kerolloz) [inline src/routes/postRouter.ts:36]:
```ts
createOne({
    ...postSchema,
})
```

is similar to 

```ts
createOne(postSchema)
```

File: src/database/prisma/schema.prisma (hunk 1 above)
Comment (MuhammedMagdyy) [reply src/database/prisma/schema.prisma:28]: We wanted to make it optional, but the product design added it to the registration process, so I made it required.
I can fix it it you want.

File: src/routes/postRouter.ts [hunk 5]
```diff
@@ -62,18 +64,33 @@ router.get(
...
+      if (!isSuccess) {
+        throw new Error();
+      }
+      photoURL = imageURL;
+    }
+    const updatePostSchema = postService.updatePostSchema.parse({
+      ...req.body,
+      id,
+      photoURL,
+    });
```
Comment (kerolloz) [inline src/routes/postRouter.ts:89]: why are you including these changes in the `Feat/user registration`

File: src/routes/userRouter.ts [hunk 6]
```diff
@@ -1,14 +1,33 @@
...
+      if (!isSuccess) {
+        throw new Error();
+      }
+      photoURL = imageURL;
+    }
+    const userScehma = userService.createUserSchema.parse({
+      ...req.body,
+      photoURL,
+    });
+    await userService.createOne(userScehma);
```
Comment (kerolloz) [inline src/routes/userRouter.ts:30]: same here.
I believe these changes should be made by @johnnabil

File: src/modules/Post/services/postService.ts (hunk 2 above)
Comment (MuhammedMagdyy) [reply src/modules/Post/services/postService.ts:22]: I wanted to make ensure that the upcoming is a type of createPostSchema

File: src/routes/userRouter.ts (hunk 6 above)
Comment (johnnabil) [reply src/routes/userRouter.ts:30]: Mohamed discovered the issue and we solved it together

File: src/routes/postRouter.ts (hunk 4 above)
Comment (MuhammedMagdyy) [reply src/routes/postRouter.ts:36]: I will fix it

File: src/routes/postRouter.ts (hunk 5 above)
Comment (MuhammedMagdyy) [reply src/routes/postRouter.ts:89]: MB, I'll delete it and assign it to john

File: src/routes/userRouter.ts (hunk 6 above)
Comment (kerolloz) [reply src/routes/userRouter.ts:30]: Yeah, but changes unrelated to `Feat/user registration` shouldn't be included here.

File: src/modules/Post/services/postService.ts (hunk 2 above)
Comment (MuhammedMagdyy) [reply src/modules/Post/services/postService.ts:22]: eslint is crying as usual :)

File: src/routes/userRouter.ts (hunk 6 above)
Comment (MuhammedMagdyy) [reply src/routes/userRouter.ts:30]: ok, I'll give the code to John to add it to the upcoming code.

**Review by kerolloz — DISMISSED** [review]

**Review by kerolloz — APPROVED** [review]

**Review by mohamadelhinamy — APPROVED** [review]
